### PR TITLE
EE-1060: Host function costs

### PR DESCRIPTION
**Pull request by mpapierski**
Ref: https://casperlabs.atlassian.net/browse/EE-1060
Researched host function values: https://github.com/CasperLabs/casperlabs-specs/blob/master/economics/transaction-fees-and-gas.md

Adds a gas charge for each host function call as according to the cost table stored in `ProtocolData`.

### CODE REVIEW

File: execution_engine/src/shared/opcode_costs.rs [hunk 1]
```diff
@@ -7,25 +7,23 @@ use serde::{Deserialize, Serialize};
...
-pub const DEFAULT_CURRENT_MEMORY_COST: u32 = 2;
-pub const DEFAULT_GROW_MEMORY_COST: u32 = 8192;
-pub const DEFAULT_REGULAR_COST: u32 = 2;
+pub const DEFAULT_BIT_COST: u32 = 300_000;
+pub const DEFAULT_ADD_COST: u32 = 210_000;
+pub const DEFAULT_MUL_COST: u32 = 240_000;
+pub const DEFAULT_DIV_COST: u32 = 320_000;
+pub const DEFAULT_LOAD_COST: u32 = 2_500_000;
+pub const DEFAULT_STORE_COST: u32 = 4_700_000;
+pub const DEFAULT_OP_CONST_COST: u32 = 110_000;
```
Comment (Fraser999) [inline execution_engine/src/shared/opcode_costs.rs:16]:
```suggestion
pub const DEFAULT_CONST_COST: u32 = 110_000;
```

File: execution_engine/src/shared/opcode_costs.rs [hunk 2]
```diff
@@ -7,25 +7,23 @@ use serde::{Deserialize, Serialize};
...
+pub const DEFAULT_LOCAL_COST: u32 = 390_000;
+pub const DEFAULT_GLOBAL_COST: u32 = 390_000;
+pub const DEFAULT_CONTROL_FLOW_COST: u32 = 440_000;
+pub const DEFAULT_INTEGER_COMPARSION_COST: u32 = 250_000;
+pub const DEFAULT_CONVERSION_COST: u32 = 420_000;
+pub const DEFAULT_UNREACHABLE_COST: u32 = 270_000;
+pub const DEFAULT_NOP_COST: u32 = 200_000; // TODO: This value is not researched
+pub const DEFAULT_CURRENT_MEMORY_COST: u32 = 290_000;
+pub const DEFAULT_GROW_MEMORY_COST: u32 = 240_000_000;
+pub const DEFAULT_REGULAR_COST: u32 = 200_000;
```
Comment (Fraser999) [inline execution_engine/src/shared/opcode_costs.rs:26]: This is `210_000` in the resources/local/chainspec.toml

File: resources/charlie/chainspec.toml [hunk 3]
```diff
@@ -115,45 +115,45 @@ gas_per_byte = 630_000_000
...
-remove_key = { cost = 61_000_000, arguments = [0, 1] }
-ret = { cost = 23_000_000, arguments = [0, 1] }
+get_system_contract = { cost = 1_100_000, arguments = [0, 0, 0] }
+has_key = { cost = 1_500_000, arguments = [0, 840] }
+is_valid_uref = { cost = 760_000, arguments = [0, 0] }
+load_named_keys = { cost = 42_000_000, arguments = [0, 0] }
+new_uref = { cost = 17_000_000, arguments = [0, 0, 590] }
+print = { cost = 20_000_000, arguments = [0, 4_600] }
+provision_contract_user_group_uref = {cost = 200_000, arguments = [0, 0, 0, 0, 0] }
+put_key = { cost = 38_000_000, arguments = [0, 840, 0, 0] }
```
Comment (Fraser999) [inline resources/charlie/chainspec.toml:143]:
```suggestion
put_key = { cost = 38_000_000, arguments = [0, 1_100, 0, 0] }
```

File: execution_engine/src/shared/host_function_costs.rs [hunk 4]
```diff
@@ -4,63 +4,151 @@ use serde::{Deserialize, Serialize};
...
+const DEFAULT_READ_VALUE_LOCAL_COST: u32 = 5_500_000;
+const DEFAULT_READ_VALUE_LOCAL_KEY_SIZE_WEIGHT: u32 = 590;
+
+const DEFAULT_REMOVE_ASSOCIATED_KEY_COST: u32 = 4_200_000;
+
+const DEFAULT_REMOVE_KEY_COST: u32 = 61_000_000;
+const DEFAULT_REMOVE_KEY_NAME_SIZE_WEIGHT: u32 = 3_200;
+
+const DEFAULT_RET_COST: u32 = 23_000_000;
+const DEFAULT_RET_VALUE_SIZE_WEIGHT: u32 = 420_000;
```
[the remark this answers is not in the document]
Comment (mpapierski) [reply execution_engine/src/shared/host_function_costs.rs:59]: True - adding a parameter with a list of urefs back to the ret call would be a bit of a hussle :)

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**